documentation/src/main/asciidoc/topics/con_performance_clustered_execution.adoc

```diff
@@ -0,0 +1,52 @@
+By executing your processing logic directly on {brandname} you spread the workload across multiple Java Virtual Machines (JVMs).
+Your code also runs in the same memory space where {brandname} stores your data, meaning that you can iterate over entries much faster.
+
+In terms of performance impact to your {brandname} deployment, that entirely depends on your code execution.
+More complex processing operations have higher performance penalties so you should approach running any code on {brandname} clusters with careful planning.
+Start out by testing your code and performing multiple execution runs on a smaller, sample data set.
+After you gather some metrics you can start identifying optimizations and understanding what performance implications of the code you're running.
+
+One definite consideration is that long running processes can start having a negative impact on normal read and write operations.
+So it is imperative that you monitor your deployment over time and continually assess performance.
+
+[discrete]
+== Embedded caches
+
+With embedded caches, {brandname} provides a `ClusterExecutor` API that lets you execute code right next to the data and directly on the cluster.
+You can execute your code on all nodes in the {brandname} or confine execution to primary owners.
+
+While the `ClusterExecutor` API lets you iterate over the entire data set, you might prefer to perform your operations on collections instead.
+To achieve this you use the `Stream` API.
+
+Another way to think about the distinction between the `ClusterExecutor` API and the `Stream` API is that clustered execution gives you processing based on {brandname} nodes and distributed streams give you processing based on data.
+
+If you want to run an operation on a single node, a group of nodes, or all nodes in a certain geographic region, then you might prefer clustered execution.
+On the other hand, if you want to run an operation that guarantees a correct result for your entire data set, then using distributed streams is a more effective option.
+
+.Cluster execution
+[source,java,options="nowrap",subs=attributes+,role="primary"]
+----
+include::code_examples/ClusterExecutor.java[]
+----
+
+.Stream
+[source,java,options="nowrap",subs=attributes+,role="secondary"]
+----
+include::code_examples/Stream.java[]
+----
+
+[discrete]
+== Remote caches
+
+For remote caches, {brandname} provides a `ServerTask` API that lets you register custom Java implementations with {brandname} Server and execute tasks programmatically by calling the `execute()` method over Hot Rod.
+
+{brandname} Server also provides scripting capabilities that are compatible with different JVM `ScriptEngine` implementations with very straightforward usage.
+All you need to do is upload your JavaScript through the {brandname} Command Line Interface (CLI) then execute it.
+// Review comment
+// Tristan: I know there is some ongoing work for the JavaScript engine. Should we reflect any of that in a statement here?
```

Comment:
@tristantarrant Hey do we need to mention anything special about the JS engine here? Should we maybe not include this? Thanks.

Reply:
The JS engine is unfortunately being dropped from Java starting with 17. I wouldn't mention anything.

Reply:
Thanks. I'll just drop this bit then.